validator/src/journal/chain_ffi.rs

```diff
@@ -0,0 +1,723 @@
+            .call_method(py, "__contains__", (block_id,), None)
+        {
+            Err(py_err) => {
+                pylogger::exception(py, "Unable to call __contains__ on BlockCache", py_err);
+                false
+            }
+            Ok(py_bool) => py_bool.extract(py).expect("Unable to extract boolean"),
+        }
+    }
+
+    fn put(&mut self, block: BlockWrapper) {
+        let gil_guard = Python::acquire_gil();
+        let py = gil_guard.python();
+
+        match self.py_block_cache
+            .set_item(py, block.header_signature(), &block)
+        {
+            Err(py_err) => {
+                pylogger::exception(py, "Unable to call __setitem__ on BlockCache", py_err);
+                ()
+            }
+            Ok(_) => (),
+        }
+    }
+
+    fn get(&self, block_id: &str) -> Option<BlockWrapper> {
+        let gil_guard = Python::acquire_gil();
+        let py = gil_guard.python();
+
+        match self.py_block_cache.get_item(py, block_id) {
+            Err(_) => {
+                // This is probably a key error, so we can return none
+                None
+            }
+            Ok(res) => Some(res.extract(py).expect("Unable to extract block")),
+        }
+    }
+}
+
+struct PyBlockValidator {
+    py_block_validator: PyObject,
+}
+
+impl PyBlockValidator {
+    fn new(py_block_validator: PyObject) -> Self {
+        PyBlockValidator { py_block_validator }
+    }
+
+    fn query_block_status(&self, fn_name: &str, block_id: &str) -> bool {
+        let gil_guard = Python::acquire_gil();
+        let py = gil_guard.python();
+
+        match self.py_block_validator
+            .call_method(py, fn_name, (block_id,), None)
+        {
+            Err(_) => {
+                // Presumably a KeyError, so no
+                false
+            }
+            Ok(py_bool) => py_bool.extract(py).expect("Unable to extract boolean"),
+        }
+    }
+}
+
+impl BlockValidator for PyBlockValidator {
+    fn in_process(&self, block_id: &str) -> bool {
+        self.query_block_status("in_process", block_id)
+    }
+
+    fn in_pending(&self, block_id: &str) -> bool {
+        self.query_block_status("in_pending", block_id)
+    }
+
+    fn validate_block(&self, block: BlockWrapper) -> Result<(), ValidationError> {
+        let gil_guard = Python::acquire_gil();
+        let py = gil_guard.python();
+
+        self.py_block_validator
+            .call_method(py, "validate_block", (block,), None)
+            .map(|_| ())
+            .map_err(|py_err| {
+                ValidationError::BlockValidationFailure(py_err.get_type(py).name(py).into_owned())
+            })
+    }
+
+    fn submit_blocks_for_verification(
+        &self,
+        blocks: &[BlockWrapper],
+        response_sender: Sender<(bool, BlockValidationResult)>,
+    ) {
+        let gil_guard = Python::acquire_gil();
+        let py = gil_guard.python();
+
+        let ctypes_module = py.import("ctypes").expect("Unable to import ctypes");
```

Comment:
Rather than do these imports in the function every time it is called, you can do them in the constructor for PyBlockValidator and then hold onto the imported modules as instance variables.